feat(FloatingPortal): allow to override FloatingPortal root node

piecyk opened the pull request:
Having structure like below, react will render to stage element. With this we would allow render `all portals` to `rootContainer` via FloatingPortalRootNodeContext.

```jsx
<body>
  <div id="before" />
  <div id="rootContainer">
    <div id="stage" />
  </div>
</body>
```

atomiks commented on packages/react/src/components/FloatingPortal.tsx:
```diff
@@ -55,13 +62,18 @@ export const useFloatingPortalNode = ({
       setPortalEl(newPortalEl);
-      const container = portalContext?.portalNode || document.body;
+
+      const container =
+        portalContext?.portalNode ??
+        floatingPortalRootNodeContext ??
+        document.body;
```
(Just to reduce transpilation cost)
```suggestion
      const container =
        portalContext?.portalNode ||
        floatingPortalRootNodeContext ||
        document.body;
```

atomiks commented on packages/react/test/unit/FloatingPortal.test.tsx:
```diff
@@ -45,3 +49,21 @@ test('empty id string does not add id attribute', () => {
+  document.body.appendChild(portalRoot);
+
+  render(
+    <FloatingPortalRootNodeContext.Provider value={document.getElementById(id)}>
```
In an SSR context, what should the user do? Use an effect and re-render?

piecyk replied:
hmm, i think so 🤔

atomiks commented:
Some thoughts:

- I think we should add a `container` prop to `FloatingPortal` and deprecate the `root` prop since it's a bit confusing when we add this.
- Should `container` and this context value should be a query selector, meaning we could also remove (deprecate for now) the `id` prop from `FloatingPortal`?
- Should we only export the `Provider`?
- Should we allow refs in addition to elements, if possible?

Basically:

- `FloatingPortalRoot` is a provider wrapping the whole app and only accepts a `root` prop. This is not necessary, since `FloatingPortal` defaults to `document.body`.

```js
<FloatingPortalRoot 
  // pass element directly
  root={document.querySelector('#root')}
  // querySelects or creates if doesn't exist
  root="#root"
>
  <Popover />
</FloatingPortalRoot>
```

- `FloatingPortal` accepts a "subroot" prop, which is the `container` prop. `id` and `root` are deprecated and will be removed in v1.

```js
function Popover() {
  return (
    <FloatingPortal 
      // pass element directly
      container={document.querySelector('#subroot')}
      // querySelects or creates if doesn't exist
      container="#subroot"
    >
      ...
    </FloatingPortal>
  );
}
```

piecyk commented:
> There's even more combinations, but to me this is really confusing to reason about, especially the a11y footguns.

Yeah, you are not alone with this one. Pushed this with removing root from FloatingPortal. 

What about `allowing direct roots` having this a11y this should not be allowed right now? Also what was the case when `id` is empty string, then we don't add set id, is this still correct?

atomiks commented:
I think we need to keep keep both `id` and `root` props for this current major to ensure backwards-compatibility, just `@deprecated` in jsdoc and removed from the documentation.

I'm also wondering in cases where the user needs to specify a root "one-off", is it impractical to do this for one-off cases?

```js
<FloatingPortalRoot root="#container">
  <FloatingPortal>
  </FloatingPortal>
</FloatingPortalRoot>
```

piecyk commented:
> I think we need to keep keep both id and root props for this current major to ensure backwards-compatibility, just @deprecated in jsdoc and removed from the documentation.

yep, good point, revert the changes, added @deprecated.

> I'm also wondering in cases where the user needs to specify a root "one-off", is it impractical to do this for one-off cases?

kinda is, but on other hand it's a corner case, imho better to keep the api simple as you mention before it can get out of hands really quickly.

atomiks commented:
Coming back to this, I think we may still want the behavior of creating a new node to be switched off for non-interactive elements to help reduce DOM pollution (even when they're tucked away inside a node inside `document.body`, and therefore collapsed in the Elements panel).


To switch off new container creation, some kind of truthy prop might work?

```js
<FloatingPortal disableContainer />
<FloatingPortal disableSubroot />
<FloatingPortal direct />
```

Also does `FloatingPortalRoot` create a node by default if `root` isn't specified? Seems like it would be a useful default.



```js
<FloatingPortalRoot>
  // creates an id-less new node inserted into `document.body`
</FloatingPortalRoot>
```

mihkeleidast commented on packages/react/test/visual/components/Drawer.tsx:
```diff
@@ -96,19 +97,21 @@ export function Drawer({children, render}: Props) {
-            content
-          ))}
-      </FloatingPortal>
+      <FloatingPortalRoot root="drawer-root">
```
I see this probably came from https://github.com/floating-ui/floating-ui/pull/2182#issuecomment-1470368315

> I'm also wondering in cases where the user needs to specify a root "one-off", is it impractical to do this for one-off cases?
> 
> ```js
> <FloatingPortalRoot root="#container">
>   <FloatingPortal>
>   </FloatingPortal>
> </FloatingPortalRoot>
> ```

IDK, API-wise it seems gross to me, I'm not a huge fan of having a context provider when it does not provide much value. IMO it should still be configurable on the portal component itself, then it would be possible to configure the default in several ways:
1. Create a wrapper component around FloatingPortal that specifies the root id and use it everywhere instead of the direct import from floating-ui
2. Use the context provider to create an implicit default (the context provider could be left to userland as well, it could be implemented via 1).

Sorry if I'm rambling, I'm not a huge fan of the context provider being the only way to configure this as a default, as I don't like the provider hell around the app, especially if this is for a very specific feature of some very specific components.

piecyk replied:
Idea here was to reduce the number of possibilities how to create portal container, like @atomiks did above with  bunch of combinations.

piecyk replied:
> Sorry if I'm rambling, I'm not a huge fan of the context provider being the only way to configure this as a default, as I don't like the provider hell around the app, especially if this is for a very specific feature of some very specific components.

No, great that you join the discussion :) I know the pain with context provider, we could keep the `root` just wrap it internally with FloatingPortalRoot 🤔

piecyk commented:
> Coming back to this, I think we may still want the behavior of creating a new node to be switched off for non-interactive elements to help reduce DOM pollution (even when they're tucked away inside a node inside document.body, and therefore collapsed in the Elements panel).

When thinking about `direct` do you also consider nesting, or in that case it should always append direct to root container?

atomiks commented:
Yeah `direct` would still allow nesting within `portalContext.portalNode`, just won't create a container with a unique id.

piecyk commented:
Ok, should be ready for review 👍

atomiks commented:
@piecyk sorry for all the reviews and back-and-forth but I'm still kinda unsure about this

I think:

- Don't have a context wrapper component as @mihkeleidast suggested
- *If* `focusManagerState.modal === false` (where the portal container must have a unique id), we provide a container or "subroot" with the id, which is the only case we do it
- Otherwise, if there's no `root` specified (defaults to `document.body`) then we create a container still but with no `id`. 
- If `root` is specified, and we don't need a subroot, then inject straight into that root
- `root` should accept a querySelector string and an HTMLElement

It's confusing me too don't worry 😓

piecyk commented:
@atomiks no worries, it's not easy issue, but also i think we are pretty close.

Regarding first point, here it boils down to ergonomic as you said before. There are two cases, first, if `all portals` should go to specific root then context wrapper is better option, without it we need to remember to add the root always on FloatingPortal or create special wrapper for it and use it rather the FloatingPortal , like `<MyFloatingPortal>stuff</MyFloatingPortal>` That's why the reason to create it. Second is that your default is document body and just for special case you would override it on fly.

> If root is specified, and we don't need a subroot, then inject straight into that root.

How we should decide if there is no need for subroot, using this explicit prop `direct`?

What about this `ensure backwards-compatibility`, should we still try to do it? or just drop the prev logic and accept new major version.

piecyk commented:
In first case

> possibly this can be made efficient by injecting everything into a single id-less container by default?

i think we can't do it, as it will break order of elements 🤔 as if we mex 1 and 2 we will have? 

```jsx
<div>
  <FloatingPortal><div id="floating1" /></FloatingPortal>
  <FloatingPortal>
    <FloatingFocusManager modal={false}>
      <div id="floating3" />
    </FloatingFocusManager>
  </FloatingPortal>
 <FloatingPortal><div id="floating2" /></FloatingPortal>
</div>
````

```jsx
<body>
  <div>
    <div id="floating1"></div>
    <div id="floating2"></div>
  </div>
  <div id="react-id"><div id="floating3"></div></div>
</body>
```

atomiks commented:
> How we should decide if there is no need for subroot, using this explicit prop direct?

This value below determines if we need a subroot with a unique id or not. However, if this value isn't `false` but there's no `root` prop specified, we would also need a subroot to avoid injecting it straight into `document.body` which may cause some performance issues.

https://github.com/floating-ui/floating-ui/blob/cfe8e3851eda4b8dff98b9e9e99c13517317f407/packages/react/src/components/FloatingPortal.tsx#L98-L99

There is a question of allowing conditionally-rendered `FloatingPortal`s, which would prevent "DOM pollution" since the portal node would only be injected while the floating element is open and not while the host component is mounted (i.e. reference element) itself. 

The reason it's not allowed to be conditionally rendered is the interaction hooks would need to wait for two render cycles after `open` becomes `true`. We might be able to work around this by checking for `floating` element's presence now instead of the `open` variable (I don't think it was possible before I added the `elements` object a couple months ago).

atomiks commented:
I've got a solution here: https://github.com/floating-ui/floating-ui/pull/2240. I've also allowed `FloatingPortal` to be conditionally rendered in another PR which is merged.

It preserves the subroot/unique id when using a custom root.

piecyk commented:
> I've also allowed FloatingPortal to be conditionally rendered in another PR which is merged.

Oooo nice 👏
